osbuild: ability to validate stage options

gicmo opened the pull request:
Add a `--validate` option to the osbuild command that will validate the options of the stages, including the build pipeline and the assembler, against the json schema of the corresponding stages.

Will abort if any errors are found and print the corresponding messages, optionally in form of json if `--json` was passed as an command line option.

I was thinking of moving the actual validation code, including the `StageInfo` class to a new `osbuild/meta.py`. Opinions?

bcl commented:
Shouldn't validation be mandatory? Or at least the default? I'm not sure if there are any cases where you wouldn't want to validate the json, wouldn't that just lead to harder to debug problems later in the pipeline execution?

gicmo commented:
> Shouldn't validation be mandatory? Or at least the default? I'm not sure if there are any cases where you wouldn't want to validate the json, wouldn't that just lead to harder to debug problems later in the pipeline execution?

Good question. We could make it mandatory but then we gain a hard dependency on `jsonschema`. Apropos, even though the import is option, pylint is not happy about that.

gicmo commented:
I wan to refactor the code to use the `ast` package to "just" parse the source of the stages, instead of actually instantiating the module.

gicmo commented:
Need to properly honour `--libdir` here as well.

bcl commented:
> I wan to refactor the code to use the `ast` package to "just" parse the source of the stages, instead of actually instantiating the module.

As long as the module import doesn't have any side-effects (which it shouldn't) why go to the extra work?

gicmo commented:
For the record errors look like this:
```
org.osbuild.fstab: 895f90b59c61d8117d38792cd5c28e14eeb1436830b20894f4daff165a13b146
ERROR: '1' is not of type 'number'

Failed validating 'type' in schema['properties']['filesystems']['items']['properties']['freq']:
    {'default': 0,
     'description': 'dump(8) period in days',
     'type': 'number'}

On instance['filesystems'][0]['freq']:
    '1'
```

dvdhrm commented:
The individual stage descriptions seem to be out-of-date. At least some tests are failing due to that. I guess that's a good thing? ;)

gicmo commented:
> Thanks for picking this up!

:heart: I am not 100% done yet: I think, all the stages should have `"additionalProperties": false` as well.

Also, the pipeline itself is not validate. @dvdhrm did the schema for it, but it is outside. What do you guys think of installing the schema as well. In `/usr/share/osbuild/schemas`, although `/usr/lib/osbuild/schemas` would be more convenient from a coding POV because `--libdir` would work for the schema as well then. Once the schemata are installed, a small tool could be crafted to export the schemata from the stages into `schemata/{stages, assemblers}/org.osbuild.<name>` and then that information could be used in the `Index` class (`get_schema(item)`). Which also means that in the future if the stages would be non-python binaries it would still work. Also, the `STAGE_DESC` could be exported to man pages. What do you guys thing?

gicmo commented:
Ok this is ready for review. There a few changes, major ones are:
  - Install the `osbuild1.json`
  - Validate the whole manifest, not only stage/assembler options
  - return value is `2` in case of validation errors
  - JSON validation result is a dict with `{"success": False, "type": "http://<id>", "errors": [...] }` (see below) Format is based on [RFC 7807](https://tools.ietf.org/rfc/rfc7807.txt)
  - New `--inspect` option which returns the description of the pipeline in JSON format, including the ids, but **not** build the pipeline. Will validate the pipeline though and return errors in JSON.
  - Fix up all invalid schemata information and pipelines


Example validation error in JSON:
```json
{
  "type": "https://osbuild.org/validation-error",
  "title": "JSON Schema validation failed",
  "success": false,
  "errors": [
    {
      "message": "Additional properties are not allowed ('foo' was unexpected)",
      "path": [
        "pipeline",
        "build",
        "pipeline"
      ]
    }
  ]
}
```

Example validation error in human readable form:
```
samples/base-qcow2.json has errors:

.pipeline.build.pipeline:
  Additional properties are not allowed ('foo' was unexpected)
```

dvdhrm commented on osbuild/meta.py:
```diff
@@ -0,0 +1,416 @@
+        for error in self._validator.iter_errors(target):
+            res += ValidationError.from_exception(error)
+
+        return res
```
Why is `_validator` a member rather than just a local variable in this function? Is there a reason to validate the schema in the constructor rather than in the `validate()` function?

gicmo replied:
A `Schema` for a given entity gets added to the cache in `Index` and thus can be re-used and thus we avoid re-creating the validator for subsequent uses. I was also going to add a `selftest` or `check` function that checks the schema itself (via the validator). Could throw in the constructor but that makes the validation code harder to read and the schemata should be validated in the CI tests so they should always be fine (the CI test should use said `check` function).

gicmo replied:
I added `Schema.check` should make that clearer.

dvdhrm commented on osbuild/meta.py:
```diff
@@ -0,0 +1,416 @@
+        self.type = klass
+
+        opts = info.get('STAGE_OPTS') or ""
+        self.info = info.get("STAGE_INFO")
```
Single-quotes? I have no clue what our guideline on that is, btw. But here it seems a bit inconsistent ;)

dvdhrm commented on osbuild/meta.py:
```diff
@@ -0,0 +1,416 @@
+        assigns = filter_type(tree.body, ast.Assign)
+        targets = [(t, a) for a in assigns for t in targets(a)]
+        info = {k: value(v) for k, v in targets if k in names}
+        return cls(klass, name, info)
```
So we now require the stages to be python scripts? If that is the case, why do all this dance, instead of just turning stages into proper python modules with well-known classes or functions that return us this information?

This is more of a general question. I am fine keeping it as it is to move forward.

gicmo replied:
This is just a pragmatic approach. I have a (stale) local branch with code that actually adds a different `main` to all stages/sources/assembler that handles `--info`, `--schema` and returns the respective information. Once that lands, adapting that function to call the "binary" and obtain the information should not be too hard.

teg replied:
General comment: Ideally we don't tie ourselves to python too much. By that I mean, if we don't assume python at all: great. If we do assume python but in a way that wouldn't be too hard to work around in the future when someone decides it is time for a rust/bash stage, I'd be perfectly fine with that too if that makes things simpler in the short turm. But I'd hesitate to tie it tightly to python by making things proper modules etc. At least that's my gut feeling.

dvdhrm commented on osbuild/meta.py:
```diff
@@ -0,0 +1,416 @@
+    """
+
+    def __init__(self, path: str):
+        self.path = path or "/usr/lib/osbuild"
```
Can we make the path mandatory? Or why is this fallback needed?

gicmo replied:
Followed `osbuild` code, but I think making this mandatory is a good idea indeed.

dvdhrm commented on osbuild/meta.py:
```diff
@@ -0,0 +1,416 @@
+    build = pipeline.get("build", {}).get("pipeline")
+    if build:
+        res = validate({"pipeline": build}, index=index)
+        result.merge(res, path=["pipeline", "build"])
```
You only need this recursion to validate the stages in each build-pipeline right? So the redundant validation could be avoided by fetching the stages in a loop and in each layer go down into the next build-pipeline? Anyway, this recursion works as well. Was just wondering.

gicmo replied:
I am not sure I understand you 100% here but I think yes ;) Could just follow down all the nested build pipelines and verify stages of those instead too.

dvdhrm commented on osbuild/meta.py:
```diff
@@ -0,0 +1,416 @@
+        for p in self.path:
+            if isinstance(p, str):
+                if " " in p:
+                    p = f"'{p}'"
```
On the other hand, `shlex.quote()` produces really ugly results if you have single-quotes in a string. Meh.

gicmo replied:
It is indeed mostly used for pretty printing. Or the other way around, it should never occur in the JSON form. I was thinking of using some heavier tool there when I wrote it. Still unsure. Happy to be swayed one way or the other.

dvdhrm commented on test/test_osbuild.py:
```diff
@@ -99,6 +99,85 @@ def load_module(base, name):
                 msg = f"Stage '{base}/{name}' has invalid STAGE_OPTS\n\t" + str(e)
```
Your commit message still says `TBD`.

gicmo commented on requirements.txt:
```diff
@@ -0,0 +1 @@
+jsonschema
```
It should indeed also be in `setup.py`! The `requirements.txt` is the (one of the?) standard python ways to install dependencies and will indeed be picked up automatically by tools, e.g. [travis](https://docs.travis-ci.com/user/languages/python/).

gicmo commented:
I reworked the `Schema` initialization that should hopefully make it clearer what is going on. Additionally, I added a `Schema.check()` function to verify the schema itself and added tests for it. I also realized that the `definitions` inside stage options need to be located at the root node of the schema otherwise it would be found. Curiously, the schema itself did pass validation without that change, but then would fail during validation.
